### DISCUSSION TITLE
Creating thumb nails in C#: slowly opening large RAW files

### DISCUSSION THREAD

**Discussion by Erik0067** [ImageMagick/ImageMagick#6857] [Development] [answered]
Hi,

I am new to Magick and I would like to use it for quickly creating small thumbnail jpg's in my C# .NET project. However, opening a large RAW file of  42 Mb (Sony ARW file) takes more than 6 seconds on my system.

Can this be done more efficiently?

Here is a piece of code:

      using (var image = new MagickImage(@"G:\Foto's\2023\2023-01-28_Nieuwe Driemanspolder\2023-01-28-1321_DSC00078.arw", MagickFormat.Arw))
      {
          var geometry = new MagickGeometry();
          geometry.IgnoreAspectRatio = false;
          geometry.Width = 200;
          image.Resize(geometry);
          image.Write("c:\\temp\\thumbNail2023-01-28-1321_DSC00078.jpg");
      }

This correctly creates a thumbnail of 200 pixels, yet the statement

new MagickImage(@"G:\Foto's\2023\2023-01-28_Nieuwe Driemanspolder\2023-01-28-1321_DSC00078.arw", MagickFormat.Arw)

takes about 6 seconds. How do I speed this up?

Some other info:
- I have installed nuget package Magick.NET-Q16-AnyCPU through Visual Studio in my project
- CPU is Intel 4770, video card NVidea GTX1650, memory is 24 Gb

Thanks in advance!

**Comment by snibgo** [comment]
What are the dimensions (width and height, in pixels) of your ARW file?

Are you running your program directly at the Windows command line, or via Docker or anything like that?

What is the text output from "`magick -version`"?

Is your disk a rotating disk, or solid-state (SSD)?

On my Windows 11 laptop, with CPU "11th Gen Intel(R) Core(TM) i7-1195G7 @ 2.90GHz 2.92 GHz", 16 GB ram, SSD, IM v7.1.1-20 Q16 with HDRI and delegate "raw" (libraw), reading a 4256x2848 pixel ARW file takes about 0.6 seconds.

I don't think that libraw uses the GPU to read raw files.

**Comment by Erik0067** [comment]
Dimensions are: BaseWidth = 7968, BaseHeight = 5320. Size of the ARW file is 42Mb, these are raw files of my SONY A7RIII camera. Yes, they are large, but programs like Faststone Image Viewer are able to display them rapidly on my system (less than 1 second!). The dimension displayed by Faststone Image Viewer is much smaller, so that probably explains it. 

And my point is, I want to create a thumbnail out of the ARW-file so I do not want to read all the details of the ARW file. After all, the final result should be a compressed thumbnail file.

My disk is a fast USB 3.0 drive. But if I move the ARW to my (very) fast SSD drive, then it still takes 6 seconds. 

I am using nuget package Magick.NET-Q16-AnyCPU in c# (visual studio),  so "magick -version" gives no results

I wonder if it possible to read the large file in such a way that the object created does not contain the sizes 7968 and 5320. I hoped an overload method of the constructor could make this happen, but I have not found that (yet?). Any more idea's?

**Comment by snibgo** [comment]
See https://www.faststone.org/FSViewerDetail.htm :

> Now it is recommended to view RAW files using the embedded preview image for faster performance (see RAW tab in the Settings)

Raw files such as ARW contain raw sensor data, and one or more JPG images the camera has made from the raw data, by de-Bayering the data, and other operations.

When IM reads a raw file, it reads the raw data and de-Bayers it, applies the camera matrix, and so on. This is much slower than simply extracting an embedded JPG preview.

If you want to extract embedded previews, I suggest using exiftool, not ImageMagick.

**Comment by dlemstra** [comment, marked as answer]
Next time you should probably ask your question here instead: https://github.com/dlemstra/Magick.NET.

And there is functionality to just read the embedded thumbnail in Magick.NET. You can use the `DngReadDefines`, set the `ReadThumbnail` to true and then `Ping` the images with those defines. After the image has been created you can extract the bytes of the thumbnail with `image.GetProfile("dng:thumbnail")`. And you should then put the array of `profile.GetData()` into the constructor of `MagickImage`.

p.s. This also works on the command line Snibgo.

**Comment by snibgo** [comment]
> p.s. This also works on the command line Snibgo.

Could you show the command line that does this? I can't see how to read the "Profile-dng:thumbnail:".

**Reply by dlemstra** [reply]
Good question. The pinged image will contain a profile with the name `dng:thumbnail` and you will need to save that to a file first. I will answer this later this week if you need more help.

**Comment by snibgo** [comment]
> I will answer this later this week if you need more help.

Yes please.

For example, using an ARW raw file:
```
C:\web\im>%IMG7%magick -ping -define dng:read-thumbnail=true DSC04420.ARW -verbose info:
```
This shows:
```
  Profiles:
    Profile-dng:thumbnail: 419534 bytes
```
... but if I try to read that:
```
C:\web\im>%IMG7%magick -ping -define dng:read-thumbnail=true DSC04420.ARW -format "%[Profile-dng:thumbnail]" info:
magick: unknown image property "%[Profile-dng:thumbnail]" @ warning/property.c/InterpretImageProperties/4234.
```

**Reply by dlemstra** [reply]
Can you share that file in a zip here?

**Comment by snibgo** [comment]
Already shared: https://github.com/ImageMagick/ImageMagick/files/12772816/DSC04420.ARW.zip

**Comment by Erik0067** [comment]
Thanks both snibgo and dlemstra!

I downloaded the commandline tool exiftool and this command gets me a thumbnail rapidly:

exiftool.exe -b -ThumbnailImage 2023-11-05-1622_DSC00394.ARW > thumbnail.jpg
and this gets me a preview:
exiftool.exe -b -PreviewImage 2023-11-05-1622_DSC00394.ARW > preview.jpg

This suits my needs as I can call exiftool.exe from my .NET program, thanks!

However, I prefer the suggestion giving me dlemstra as it is implemented in a nuget package.
After some googling I found out that you already answered my question in the .NET forum (I will use that forum from now on), see https://github.com/dlemstra/Magick.NET/discussions/1006

I modified the code provided by you in this discussion a bit:

           string FileName = "G:\\Foto's\\2023\\2023-01-28_Nieuwe Driemanspolder\\2023-01-28-1321_DSC00078.arw";
           using (var image = new MagickImage())
           {
               image.Settings.SetDefines(new DngReadDefines
               {
                   ReadThumbnail = true
               });

               image.Ping(FileName);

               var profile = image.GetProfile("dng:thumbnail");
               File.WriteAllBytes(@"c:\temp\thumbnail.jpg", profile.GetData());
           }

This works super fast! One thing strikes me though. The dimension of the created file c:\temp\thumbnail.jpg is 1616 * 1080 pixels, that is the size of the preview image of exiftool. Exiftool is also able to extract a small thumbnail of 160 * 120 pixels. Can you generate such a small file as well by using different settings (if so, which)? Or do I have to resize it to 160 * 120 pixels for a small thumbnail?

**Reply by dlemstra** [reply]
You will need to read and resize that thumbnail image. I have also added an example about this to the documentation of Magick.NET
